File: 232.implement-queue-using-stacks.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=232 lang=cpp
 *
 * [232] Implement Queue using Stacks
 */

// @lc code=start
class MyQueue
{
private:
    stack<int> in;
    stack<int> out;

public:
    MyQueue()
    {
    }

    void push(int x)
    {
        in.push(x);
    }

    int pop()
    {
        if (!out.empty())
        {
            int x = out.top();
            out.pop();
            return x;
        }
        while (!in.empty())
        {
            out.push(in.top());
            in.pop();
        }
        int x = out.top();
        out.pop();
        return x;
    }

    int peek()
    {
        if (!out.empty())
        {
            return out.top();
        }
        while (!in.empty())
        {
            out.push(in.top());
            in.pop();
        }
        return out.top();
    }

    bool empty()
    {
        return in.empty() && out.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */
// @lc code=end

File: Java/Grind75/Week1/232.implement-queue-using-stacks.java
package Grind75.Week1;

import java.util.*;
/*
 * @lc app=leetcode id=232 lang=java
 *
 * [232] Implement Queue using Stacks
 */

// @lc code=start
class MyQueue {
    private Stack<Integer> stack1;
    private Stack<Integer> stack2;

    public MyQueue() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }

    public void push(int x) {
        stack1.push(x);
    }

    public int pop() {
        if (!stack2.isEmpty())
            return stack2.pop();
        while (!stack1.empty())
            stack2.push(stack1.pop());
        return stack2.pop();
    }

    public int peek() {
        if (!stack2.empty())
            return stack2.peek();
        while (!stack1.empty())
            stack2.push(stack1.pop());
        return stack2.peek();
    }

    public boolean empty() {
        return stack1.empty() && stack2.empty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */
// @lc code=end
